client/lock.go

```diff
@@ -0,0 +1,333 @@
+package client
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/google/uuid"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+	"github.com/xline-kv/go-xline/xlog"
+	"go.uber.org/zap"
+)
+
+// Client for Lock operations.
+type lockClient struct {
+	// Name of the LockClient
+	name string
+	// The client running the CURP protocol, communicate with all servers.
+	curpClient curpClient
+	// The lease client
+	leaseClient leaseClient
+	// The watch client
+	watchClient watchClient
+	// Auth token
+	token string
+}
+
+// Creates a new `LockClient`
+func newLockClient(
+	name string,
+	curpClient curpClient,
+	innerLeaseClient leaseClient,
+	innerWatchClient watchClient,
+	token string,
+) lockClient {
+	return lockClient{
+		name:        name,
+		curpClient:  curpClient,
+		leaseClient: innerLeaseClient,
+		watchClient: innerWatchClient,
+		token:       token,
+	}
+}
+
+// The inner lock logic
+func (c *lockClient) lockInner(
+	prefix string,
+	key string,
+	leaseId int64,
+	lockSuccess bool,
+) (*xlineapi.LockResponse, error) {
+	logger := xlog.GetLogger()
+
+	txn := c.createAcquireTxn(prefix, leaseId)
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_TxnRequest{
+			TxnRequest: &txn,
+		},
+	}
+	proposeId := c.generateProposeId()
+	cmd := xlineapi.Command{
+		Request:   &requestWithToken,
+		ProposeId: proposeId,
+	}
+	res, err := c.curpClient.propose(&cmd, false)
+	if err != nil {
+		return nil, err
+	}
+	txnRes := res.CommandResp.GetTxnResponse()
+
+	if res.SyncResp == nil {
+		return nil, errors.New("SyncRes always has value when use slow path")
+	}
+	myRev := res.SyncResp.Revision
+
+	ownerRes := txnRes.Responses[1].GetResponseRange()
+	if ownerRes == nil {
+		return nil, errors.New("get owner res fail")
+	}
+	ownerKey := ownerRes.Kvs
+
+	var header *xlineapi.ResponseHeader
+	if len(ownerKey) > 0 && ownerKey[0] != nil && ownerKey[0].CreateRevision == myRev {
+		header = ownerRes.Header
+	} else {
+		c.waitDelete(prefix, myRev)
+		rangeReq := xlineapi.RangeRequest{
+			Key: []byte(key),
+		}
+		requestWithToken := xlineapi.RequestWithToken{
+			Token: &c.token,
+			RequestWrapper: &xlineapi.RequestWithToken_RangeRequest{
+				RangeRequest: &rangeReq,
+			},
+		}
+		proposeId := c.generateProposeId()
+		cmd := xlineapi.Command{Request: &requestWithToken, ProposeId: proposeId}
+
+		res, err := c.curpClient.propose(&cmd, true)
+		if err == nil {
+			res := res.CommandResp.GetRangeResponse()
+			if len(res.Kvs) == 0 {
+				return nil, errors.New("Rpc error session expired")
+			}
+			header = res.Header
+		} else {
+			_, err = c.deleteKey([]byte(key))
+			if err != nil {
+				logger.Warn("", zap.Error(err))
```

Comment:
please add some hint message for this error log

Reply:
Done.